libraries/chain/block_schedule.cpp

```diff
@@ -0,0 +1,292 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+#include <eos/chain/block_schedule.hpp>
+#include <eos/chain/block.hpp>
+
+namespace eos { namespace chain {
+
+static uint next_power_of_two(uint input) {
+   if (input == 0) {
+      return 0;
+   }
+
+   uint result = input;
+   result--;
+   result |= result >> 1;
+   result |= result >> 2;
+   result |= result >> 4;
+   result |= result >> 8;
+   result |= result >> 16;
+   result++;
+   return result;
+};
+
+typedef std::hash<decltype(AccountName::value)> account_hash;
+static account_hash account_hasher;
+
+struct schedule_entry {
+   schedule_entry(uint _cycle, uint _thread, pending_transaction const * _transaction) 
+      : cycle(_cycle)
+      , thread(_thread)
+      , transaction(_transaction) 
+   {}
+
+   uint cycle;
+   uint thread;
+   pending_transaction const *transaction;
+
+   friend bool operator<( schedule_entry const &l, schedule_entry const &r ) {
+      if (l.cycle < r.cycle) {
+            return true;
+      } else if (l.cycle == r.cycle) {
+            return l.thread < r.thread;
+      } 
+
+      return false;
+   }
+};
+
+static block_schedule from_entries(vector<schedule_entry> &entries) {
+   // sort in reverse to save allocations, this should put the highest thread index
+   // for the highest cycle index first meaning the naive resize in the loop below
+   // is usually the largest and only resize
+   auto reverse = [](schedule_entry const &l, schedule_entry const &r) {
+      return !(l < r);
+   };
+
+   std::sort(entries.begin(), entries.end(), reverse);
+
+   block_schedule result;
+   for(auto const & entry : entries) {
+      if (result.cycles.size() <= entry.cycle) {
+         result.cycles.resize(entry.cycle + 1);
+      }
+
+      auto &cycle = result.cycles.at(entry.cycle);
+
+      if (cycle.size() <= entry.thread) {
+         cycle.resize(entry.thread + 1);
+      }
+
+      // because we are traversing the schedule in reverse to save
+      // allocations, we cannot emplace_back as that would reverse
+      // the transactions in a thread
+      auto &thread = cycle.at(entry.thread);
+      thread.transactions.emplace(thread.transactions.begin(), *entry.transaction);
+   }
+
+   return result;
+}
+
+template<typename CONTAINER>
+auto initialize_pointer_vector(CONTAINER const &c) {
```

Comment:
stylistic,   const CONTAINER& c, once again the '&' is part of the type, not the variable.